README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# pufr <a href="https://servinagrero.github.io/pufr/"><img src="man/figures/logo.svg" align="right" height="139" /></a>

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![Codecov test
coverage](https://codecov.io/gh/servinagrero/pufr/branch/main/graph/badge.svg)](https://app.codecov.io/gh/servinagrero/pufr?branch=main)
<!-- badges: end -->

## Overview

Physical Unclonable Functions, commonly knows as PUFs, are a mechanism
that leverages the manufacture variability of devices to generate
secrets. PUFs can be used to identify devices or to generate secrets.
This package provides functions and algorithms for their analysis and
evaluation. The functions provided in this package are optimised to be
fast enough for most use cases.

A Python implementation can be found in
[Github](https://github.com/servinagrero/pufrpy).

You can learn more about PUFs and their evaluation in
`vignette("pufr")`.

## Installation

You can install the development version of pufr from
[GitHub](https://github.com/servinagrero/pufr) with:

``` r
# install.packages("devtools")
devtools::install_github("servinagrero/pufr")
```

## Basic functionality

PUFR provides functions and operators to work both with bit vectors
(vectors containing 1s and 0s) and CRPs matrices (2D or 3D matrices with
1s and 0s).

``` r
library(pufr)

# rbits can be used to generate bits with desired probabilities
v1 <- rbits(50)
v2 <- rbits(50)

hamming_dist(v1, v2)
#> [1] 19
# Or alternatively
v1 %HD% v2
#> [1] 19

# Shannon entropy for bit vectors
entropy_bits(v1)
#> [1] 0.9895875
```

## Metrics

PUFR provides utilities to calculate the canonical metrics (uniformity,
bitaliasing, uniqueness and reliability) of CRPs. They can be calculated
using their respective functions.

``` r
# 5 devices with 10 challenges each
crps <- rbits(c(5, 10))

uniformity(crps)
#> [1] 0.5 0.5 0.4 0.6 0.4
bitaliasing(crps)
#>  [1] 0.4 0.6 0.8 0.6 0.4 0.4 0.2 0.6 0.2 0.6
uniqueness(crps)
#>  [1] 0.6 0.3 0.5 0.1 0.5 0.7 0.5 0.2 0.8 0.4

# For a 2D matrix, each column contains a sample
# In a 3D matrix, the 3rd dimension represents the samples
# 5 samples of 5 devices with 10 challenges each
crps <- rbits(c(5, 10, 5))
reliability(crps)
#>      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
#> [1,] 1.00 0.75 0.25 0.75 0.25 0.25 0.25 0.75 0.00  0.50
#> [2,] 0.75 0.75 0.25 0.50 1.00 0.50 0.50 0.75 0.25  0.25
#> [3,] 0.50 0.75 0.25 0.75 1.00 0.25 0.50 0.50 1.00  0.75
#> [4,] 0.25 0.50 0.25 0.75 0.25 0.50 0.50 0.50 0.75  0.25
#> [5,] 0.50 0.50 0.50 0.25 0.25 0.50 0.75 0.25 0.50  0.25
```

All the metrics can be calculated at once by using the `metrics`
function, which handles both 2D and 3D matrices.

``` r
crps <- rbits(c(5, 10))
met <- metrics(crps)
```

### Exploration of metrics

Metrics created through the `metrics` function can be plotted and
summarised easily.

In the case of a 2D CRP matrix, 3 histograms are created, for
uniformity, bitaliasing and uniqueness respectively.

``` r
crps <- rbits(c(10, 100))
plot(metrics(crps))
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />

In the case of a 3D CRP matrix, response samples are drawn from the 3rd
dimension. The histograms are now grouped per sample and an additional
raster plot is created to show the reliability of each independent
response.

``` r
crps <- rbits(c(10, 100, 5))
plot(metrics(crps))
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" />


NEWS.md
# pufr 0.5.1

* General fixes
* Improved the test suite. This test suite will be used to check the proper behaviour of all the functions and will be used as the reference for the Python version.
* Store names of the crps (devices, challenges and samples) when computing metrics

# pufr 0.5.0

* Make `rbits` create vector, matrix and array, for convenience purposes.
* Added `pufmetrics` class. Allows for automatic plotting and exploration of all metrics.
* Added `compare_pairwise` function to perform functions on all pairs of devices.
* Removed parallel code. If needed, parallelization should be implemented by the user.
* Removed `crps_weight` function in favor of `uniformity` and `bitaliasing`

# pufr 0.4.0

* Renamed `crps_uniqueness` to `uniqueness`.
* Add `entropy_shannon`, `entropy_bits` and `entropy_p`.
* Add `compare_pairwise` to compare a matrix by pairs of rows.
* Make `intra_hd` accept a vector, a matrix and an array.
* Add `reliability` to compute the complementary of `intra_hd`.

# pufr 0.3.0

* `crps_uniqueness` now returns a list of values instead of the average uniqueness. (#1)
* `intra_hd` returns a 2D matrix if the argument supplied is a 3D array.
* Added `%<>%` and `%</>%` operators for hamming distance.

# pufr 0.2.1

* Add `register_parallel` function to register a parallel cluster.

# pufr 0.2.0

* Add automatic detection of the `parallel` package to parallelize the computations.
* Renamed `uniqueness` to `crps_uniqueness`.
* Renamed `entropy_lgl` to `entropy_bits`.
* Add unit test suite.

# pufr 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(pufr)

test_check("pufr")


tests/testthat/test-utils.R
test_that("rbits generates the correct size", {
  v <- rbits(10)
  m <- rbits(c(5, 10))
  arr <- rbits(c(5, 10, 3))

  expect_equal(length(v), 10)
  expect_equal(dim(m), c(5, 10))
  expect_equal(dim(arr), c(5, 10, 3))
})

test_that("rbits fails if number of probs does not match size", {
  P <- runif(10)
  expect_error(rbits(15, P))
})

test_that("hamming_weight handles `NA`", {
  v <- sample(c(0, 1, NA), 1000, replace = TRUE)
  weight <- hamming_weight(v, norm = TRUE)
  expect_lt(abs(weight - 0.5), 0.1)
})

test_that("hamming_dist fails on different lengths", {
  expect_error(hamming_dist(c(0, 0), c(0, 1, 0)))
})

test_that("ratio_bits discards `NA`", {
  ratio <- ratio_bits(c(1, 0, 1, NA, NA))
  expect_equal(ratio, 1 / 3)
})

test_that("compare_pairwise creates all pairs of rows", {
  m <- rbits(c(5, 10))
  npairs <- (nrow(m) * (nrow(m) - 1)) / 2
  res <- compare_pairwise(m, sum)
  expect_equal(length(res), npairs)
})

test_that("equal_to_idx works properly", {
  x <- rbits(100)
  res <- equal_to_idx(x)
  expect_lt(abs(0.5 - mean(res)), 0.1)

  x <- rep(1, 100)
  expect_true(all(equal_to_idx(x) == 1))

  expect_error(equal_to_idx(rbits(5), 10))
  expect_error(equal_to_idx(rbits(5), -2))
})

test_that("crps_to_df works properly", {
  arr <- rbits(c(5, 20, 3))
  df <- crps_to_df(arr)
  expect_equal(length(unique(df$device)), dim(arr)[1])
  expect_equal(length(unique(df$challenge)), dim(arr)[2])
  expect_equal(length(unique(df$sample)), dim(arr)[3])

  bits <- unlist(lapply(seq_len(dim(arr)[3]), function(s) as.vector(t(arr[, , s]))))
  expect_true(is.data.frame(df))
  expect_equal(df$response, bits)
})

test_that("conversion from df to crps is reversible", {
  arr <- rbits(c(5, 20, 3))
  res <- df_to_crps(crps_to_df(arr))
  expect_equal(dim(res), dim(arr))
  expect_true(all(arr == res))
})


tests/testthat/test-metrics.R
test_that("metrics has the correct dimensions", {
  m <- rbits(c(5, 10))
  expect_equal(dim(metrics(m)), c(5, 10, 1))
  arr <- rbits(c(5, 10, 3))
  expect_equal(dim(metrics(arr)), c(5, 10, 3))
})


test_that("reliability is NA when a matrix is supplied", {
  arr <- rbits(c(5, 10, 3))
  met <- metrics(arr[,,1])
  expect_true(is.na(met$reliability))
  met <- metrics(arr)
  expect_true(is.matrix(met$reliability))
})

test_that("with_entropy modifies uniformity and bitaliasing", {
  m <- rbits(c(5, 10), p = 1)
  met <- metrics(m)
  expect_true(all(met$uniformity == 1))
  expect_true(all(met$bitaliasing == 1))

  met_ent <- with_entropy(met)
  expect_true(all(met_ent$uniformity == 0))
  expect_true(all(met_ent$bitaliasing == 0))
})


R/metrics.R
library(ggplot2)
library(dplyr)
library(cowplot)
library(viridis)

#' Compute the quality metrics of the given CRP matrix
#'
#' @param crps A 2D or 3D CRP matrix
#' @param with_entropy If `TRUE`, wrap the results of the metrics with the Shannon entropy. By default `FALSE`.
#'
#' @returns The calculated metrics in a `pufmetrics` object.
#'
#' @export
#' @examples
#' crps <- rbits(c(5, 50))
#' metrics(crps)
#'
#' crps <- rbits(c(5, 50, 3))
#' metrics(crps)
metrics <- function(crps, with_entropy = FALSE) {
  m <- structure(list(reliability = NA), class = "pufmetrics")

  m$devices <- `if`(is.null(rownames(crps)), seq_len(nrow(crps)), rownames(crps))
  m$challenges <- `if`(is.null(colnames(crps)), seq_len(ncol(crps)), colnames(crps))

  if (is.matrix(crps)) {
    m$uniformity <- as.vector(pufr::uniformity(crps))
    m$bitaliasing <- as.vector(pufr::bitaliasing(crps))
    m$uniqueness <- pufr::uniqueness(crps)
  } else {
    m$samples <- `if`(is.null(dimnames(crps)[3]), seq_len(dim(crps)[3]), dimnames(crps)[[3]])
    m$uniformity <- lapply(m$samples, function(s) as.vector(uniformity(crps[, , s])))
    m$bitaliasing <- lapply(m$samples, function(s) as.vector(bitaliasing(crps[, , s])))
    m$uniqueness <- lapply(m$samples, function(s) as.vector(uniqueness(crps[, , s])))
    m$reliability <- pufr::reliability(crps)
  }
  return(`if`(with_entropy, with_entropy(m), m))
}


#' Convert metrics to entropy calculation
#'
#' @param metrics A pufmetrics object
#'
#' @returns The new pufmetrics object where the metrics have been passed to `entropy_p` description
#'
#' @seealso [entropy_p][pufr::entropy_p]
#' @export
#' @examples
#' crps <- rbits(c(5, 50))
#' m <- metrics(crps)
#' m
#' with_entropy(m)
with_entropy <- function(metrics) {
  stopifnot(class(metrics) == "pufmetrics")

  m <- metrics

  if (dim(metrics)[3] == 1) {
    m$uniformity <- entropy_p(metrics$uniformity)
    m$bitaliasing <- entropy_p(metrics$bitaliasing)
  } else {
    m$uniformity <- lapply(metrics$uniformity, entropy_p)
    m$bitaliasing <- lapply(metrics$bitaliasing, entropy_p)
  }
  return(m)
}

#' Create a plot with a summary of all the metrics
#'
#' If the metrics come from a 2D CRP table, 3 histograms are created for uniformity, bitaliasing and uniqueness.
#'
#' In the case of a 3D CRP table, 4 graphs are created, 3 histograms for uniformity, bitaliasing and uniqueness, combining all samples, and a matrix showing the reliability of each response.
#'
#' @param x The PUF metrics
#' @param ... Additional parameters
#'
#' @export
#' @examples
#' ## With a single sample
#' crps <- rbits(c(5, 50))
#' plot(metrics(crps))
#'
#' ## With multiple samples
#' crps <- rbits(c(5, 50, 3))
#' plot(metrics(crps))
plot.pufmetrics <- function(x, ...) {
  histogram <- function(x) {
    df <- do.call(rbind, x) %>%
      reshape2::melt() %>%
      rename(sample = "Var1")
    ggplot(df) +
      geom_histogram(
        aes(x = value, group = sample, fill = as.factor(sample)),
        position = "dodge", bins = 10
      ) +
      scale_fill_viridis_d() +
      scale_x_continuous(n.breaks = 10) +
      coord_cartesian(xlim = c(0, 1)) +
      labs(fill = "Sample")
  }

  if (is.matrix(x$reliability)) {
    p_unif <- histogram(x$uniformity)
    p_ba <- histogram(x$bitaliasing)
    p_uniq <- histogram(x$uniqueness)

    p_rel <- reshape2::melt(x$reliability) %>%
      ggplot() +
      geom_raster(aes(x = Var2, y = Var1, fill = value)) +
      scale_fill_viridis() +
      labs(x = "Challenge", y = "Device", fill = "Rel")

    cowplot::plot_grid(
      p_unif,
      p_ba,
      p_uniq,
      p_rel,
      labels = c("Uniformity", "Bitaliasing", "Uniqueness", "Reliability"),
      ncol = 2, nrow = 2
    )
  } else {
    plist <- lapply(
      x[c("uniformity", "bitaliasing", "uniqueness")],
      function(mat) {
        data.frame(value = mat) %>%
          ggplot() +
          geom_histogram(aes(x = value), bins = 10) +
          scale_x_continuous(n.breaks = 10) +
          coord_cartesian(xlim = c(0, 1))
      }
    )

    cowplot::plot_grid(
      plotlist = plist,
      labels = c("Uniformity", "Bitaliasing", "Uniqueness"),
      ncol = 1
    )
  }
}

#' Returns the dimensions of the metrics
#'
#' The dimensions are defined as the number of devices, number of challenges and number of samples. When only 1 sample has been used, the 3rd dimension defaults to 1.
#'
#' @param x The PUF metrics
#'
#' @returns The number of devices, challenges and samples from the metrics
#'
#' @export
#' @examples
#' crps <- rbits(c(5, 50))
#' dim(metrics(crps))
#'
#' crps <- rbits(c(5, 50, 3))
#' dim(metrics(crps))
dim.pufmetrics <- function(x) {
  if (is.matrix(x$reliability)) {
    c(nrow(x$reliability), ncol(x$reliability), length(x$uniformity))
  } else {
    c(length(x$uniformity), length(x$bitaliasing), 1)
  }
}


#' Returns the length of the metrics
#'
#' The length of the metrics corresponds to the total number of responses
#'
#' @param x The PUF metrics
#'
#' @returns The total number of responses
#'
#' @export
#' @examples
#' crps <- rbits(c(5, 10))
#' length(metrics(crps))
#'
#' crps <- rbits(c(5, 10, 3))
#' length(metrics(crps))
length.pufmetrics <- function(x) {
  prod(dim(x))
}


#' Create a summary of the metrics
#'
#' @param object The PUF metrics
#' @param ... Rest of the parameters
#'
#' @returns The total number of responses
#'
#' @export
#' @examples
#' arr <- rbits(c(5, 10, 3))
#' summary(metrics(arr))
summary.pufmetrics <- function(object, ...) {
  pad <- function(x, width = 15) stringr::str_pad(x, width = width, side = "right")
  dims <- sprintf("%3d", dim(object))
  cat(pad("Number of devices:", 22), dims[1], "\n")
  cat(pad("Number of challenges:", 22), dims[2], "\n")
  cat(pad("Number of samples:", 22), dims[3], "\n\n")

  header <- str_pad(c("Min", "Mean", "SD", "Max"), width = 5, side = "left")
  cat("               ", header, "\n")
  mysum <- function(name, x) {
    x <- `if`(is.list(x), x[[1]], x)
    vals <- sprintf("%#.3f", c(min(x), mean(x), sd(x), max(x)))
    cat(pad(name), vals, "\n")
  }

  mysum("Uniformity:", object$uniformity)
  mysum("Bitaliasing:", object$bitaliasing)
  mysum("Uniqueness:", object$uniqueness)
  if (is.matrix(object$reliability)) {
    mysum("Reliability:", object$reliability)
  }
}


R/formulas.R
## usethis namespace: start
#' @useDynLib pufr, .registration = TRUE
#' @importFrom Rcpp sourceCpp
#' @importFrom cowplot plot_grid
#' @import ggplot2
#' @import dplyr
#' @import viridis
#' @importFrom stringr str_pad
#' @importFrom stats rbinom
#' @importFrom stats sd
#' @importFrom reshape2 melt
#' @importFrom abind abind
## usethis namespace: end
NULL

#' Shannon entropy for numeric values
#'
#' @description
#' The Shannon entropy of a vector is calculated as:
#' \deqn{H(X) := -\sum_{x \in \chi} p(x) \log p(x) = E[-\log p(x)]}
#'
#' @details
#' When calculating the probabilities, it is assumed by convention that \eqn{(0\cdot \log 0 = 0)} and \eqn{(1 \cdot \log 1 = 1)}.
#'
#' @param v A vector of values
#'
#' @returns The Shannon entropy of the vector
#'
#' @export
#' @examples
#' v <- sample(1:10, 50, TRUE, rep(0.1, 10))
#' entropy_shannon(v)
entropy_shannon <- function(v) {
  freqs <- table(v) / length(v)
  -sum(freqs * log2(freqs))
}

#' Shannon entropy for binary vectors
#'
#' @details
#' The probability \eqn{p(1)} corresponds to the normalized [hamming_weight] of the vector.
#'
#' @param v A binary vector
#'
#' @returns The Shannon entropy of the vector
#'
#' @export
#' @seealso [entropy_shannon][pufr::entropy_shannon]
#' @examples
#' entropy_bits(c(0, 0, 0))
#' entropy_bits(c(1, 1, 1))
#' entropy_bits(rbits(20))
entropy_bits <- function(v) {
  p_ones <- hamming_weight(v, norm = TRUE)
  entropy_p(p_ones)
}

#' Shannon entropy for probabilities
#'
#' @param v A vector of probabilities where each value is treated as the probability of success or P(1)
#'
#' @returns The Shannon entropy of each probability
#'
#' @seealso [entropy_shannon][pufr::entropy_shannon]
#' @export
entropy_p <- function(v) {
  p_zero <- 1 - v
  vals <- -((v * log2(v)) + (p_zero * log2(p_zero)))
  ifelse(is.nan(vals), 0, vals)
}

#' Uniformity of a bit vector or a CRP matrix
#'
#' This function assumes that if the CRPs are supplied in a 2D matrix, each row corresponds to a device and each column corresponds to a CRP.
#'
#' @details
#' Uniformity measures the distribution of 1s and 0s across all CRPs for each device. To calculate uniformity, `margin` should be 1.
#'
#' \deqn{Uniformity(d) = \frac{1}{\#C} \sum_{r \in R_d} r}
#'
#' @param crps A bit vector or a 2D CRP matrix
#'
#' @returns If `crps` is a vector, the normalized Hamming weight. If `crps` is a matrix, the normalized Hamming weight of each row.
#'
#' @export
#' @seealso [hamming_weight][pufr::hamming_weight]
#' @examples
#' ## Uniformity of a vector
#' uniformity(rbits(50))
#'
#' ## Uniformity of a matrix
#' uniformity(rbits(c(10, 50)))
uniformity <- function(crps) {
  stopifnot(is.vector(crps) || is.matrix(crps))
  if (is.vector(crps)) {
    hamming_weight(crps, norm = TRUE)
  } else if (is.matrix(crps)) {
    apply(crps, 1, hamming_weight, norm = TRUE)
  }
}

#' Bitaliasing of a CRP matrix
#'
#' Unlike `Uniformity`, it does not make sense to calculate the bitaliasing of a single bit vector. Thus, `bitaliasing` only works on 2D matrix.
#'
#' Like `Unformity`, this function assumes that the CRPs are supplied in a 2D matrix, where each row corresponds to a device and each column corresponds to a CRP.
#'
#' @details
#' Bitaliasing measures the distribution of 1s and 0s for a single CRPs across all devices.
#'
#' \deqn{Bitaliasing(c) = \frac{1}{\#D} \sum_{d \in D} r_c}
#'
#' @param crps A bit vector or a 2D CRP matrix
#'
#' @returns The normalized Hamming weight of each column.
#'
#' @export
#' @seealso [hamming_weight][pufr::hamming_weight]
#' @examples
#' ## Bitaliasing of a matrix
#' bitaliasing(rbits(c(5, 50)))
bitaliasing <- function(crps) {
  stopifnot(is.matrix(crps))
  apply(crps, 2, hamming_weight, norm = TRUE)
}


#' Compare all values in a vector to the reference index
#'
#' @param x A numeric or binary vector
#' @param ref The index of the value used as reference. By default `1`.
#'
#' @returns A numeric vector containing 1 in the positions with equal values to the reference and 0 otherwise.
#' @export
#' @examples
#' v <- rbits(10)
#' v
#' equal_to_idx(v, 1)
equal_to_idx <- function(x, ref = 1) {
  stopifnot("ref is outside the range" = ref >= 1 & ref <= length(x))
  as.numeric(x == x[ref])[-ref]
}

#' Intra Hamming distance of CRPs
#'
#' @description
#' The intra Hamming distance of two sets of CRPs measures the numbers of CRPs that differ. N sets of CRPs are equal if their intra Hamming distances are 0.
#'
#' A response is reliable if it does not change in time, thus, its intra Hamming distance is equal to 0.
#'
#' If `crps` is a 2D matrix, it is assumed that each row corresponds to a sample and each column to a CRP. The `ref_sample` will dictate which row is the sample taken as reference. In the case of a 3D matrix, each row corresponds to a device, each column to a CRP and the 3rd dimension represents the different samples.
#'
#' @details
#' The order of the samples is calculated as \code{setdiff(seq_len(nsamples), ref_sample)} where `nsamples` corresponds to \code{nrow(crps)} in the case of a 2D matrix and \code{dim(crps)[3]} in the case of a 3D matrix.
#'
#' @param crps A binary vector, 2D matrix or 3D array.
#' @param ref Numeric index for the reference sample: If `crps` is a vector, is the index of the reference sample; If `crps` is a 2D matrix, the row to use as reference; If `crps` is a 3D array, the row for all 3rd dimension matrix.
#'
#' @returns If `crps` is a vector, the intra Hamming distance of the vector. If `crps` is a 2D matrix, the reliability of each column as a vector of size \code{ncol(crps) - 1}. If `crps` is a 3D array, a 2D matrix where each row contains the intra Hamming distance all samples.
#'
#' TODO: Maybe return the list of comparison to calculate the mean and sd
#'
#' @export
#' @seealso [hamming_dist][pufr::hamming_dist]
#'
#' @examples
#' ## Bit values
#' v <- c(1, 0, 0, 1, 1)
#' intra_hd(v, 1)
#'
#' ## Set of CRPs
#' mat <- rbits(c(5, 10))
#' intra_hd(mat, 1)
#'
#' ## Set of devices with their respective samples
#' mat <- rbits(c(5, 10, 3))
#' intra_hd(mat)
intra_hd <- function(crps, ref = 1) {
  if (is.vector(crps)) {
    return(equal_to_idx(crps, ref))
  }

  if (is.matrix(crps)) {
    return(apply(crps, 2, function(x) equal_to_idx(x, ref)))
  }

  if (is.array(crps)) {
    return(apply(crps, c(1, 2), function(x) mean(equal_to_idx(x, ref))))
  }

  return(1)
}

#' Reliability of CRPs
#'
#' The reliability of a PUF is defined as \eqn{1 - Intra_{HD}}
#' @inheritParams intra_hd
#' @export
#' @seealso [intra_hd][pufr::intra_hd]
#'
#' @examples
#' ## Set of CRPs
#' mat <- rbits(c(5, 10))
#' intra <- intra_hd(mat, 1)
#' all(1 - intra == reliability(mat, 1))
reliability <- function(crps, ref = 1) {
  1 - intra_hd(crps, ref)
}

#' @rdname uniqueness
#' @export
inter_hd <- uniqueness


src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <Rcpp.h>

using namespace Rcpp;

#ifdef RCPP_USE_GLOBAL_ROSTREAM
Rcpp::Rostream<true>&  Rcpp::Rcout = Rcpp::Rcpp_cout_get();
Rcpp::Rostream<false>& Rcpp::Rcerr = Rcpp::Rcpp_cerr_get();
#endif

// uniqueness
NumericVector uniqueness(const NumericMatrix& crps);
RcppExport SEXP _pufr_uniqueness(SEXP crpsSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< const NumericMatrix& >::type crps(crpsSEXP);
    rcpp_result_gen = Rcpp::wrap(uniqueness(crps));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_pufr_uniqueness", (DL_FUNC) &_pufr_uniqueness, 1},
    {NULL, NULL, 0}
};

RcppExport void R_init_pufr(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


R/utils.R
#' Random bit vector, matrix or array
#'
#' This function is a wrapper around [rbinom] to generate bit vectors. The matrix and array version are creating by row.
#'
#' @param size The size of the vector. Can be a list of dimensions to create a vector, matrix or array. If a vector larger than 3 is provided, each value is treated as the probability of obtaining 1 and a vector of bits is generated using a binomial distribution.
#' @param p Probability of obtaining a 1. By default it's `0.5`.
#' @param ... Extra arguments passed to matrix or array
#'
#' @returns The generated bit vector
#' @export
#'
#' @examples
#' ## Unbiased probabilities
#' rbits(10)
#'
#' ## Biased probabilities
#' rbits(10, p = 0.8)
#'
#' ## Matrix of bits
#' rbits(c(3, 4))
#'
#' ## 3D Array of bits
#' rbits(c(3, 4, 2))
#'
#' ## Individual probabilities
#' rbits(c(3, 3), runif(9, max = 0.5))
rbits <- function(size, p = 0.5, ...) {
  if (length(p) > 1 && prod(size) != length(p)) {
    stop("Number of probabilities does not match number of bits")
  }

  bits <- rbinom(prod(size), size = 1, prob = p)

  if (length(size) == 1) {
    bits
  } else if (length(size) == 2) {
    matrix(bits, nrow = size[[1]], ncol = size[[2]], byrow = TRUE, ...)
  } else if (length(size) == 3) {
    aperm(array(bits, dim = c(size[2], size[1], size[3]), ...), c(2, 1, 3))
  }
}

#' Hamming distance of two vectors
#'
#' @description
#' The Hamming distance of two vectors corresponds to the number of positions where the values differ.
#'
#' @details
#' NAs are discarded in any of the vectors by using `na.rm = TRUE` in `sum`.
#'
#' @param x A numeric or logical vector
#' @param y A numeric or logical vector
#' @param norm If `TRUE` normalize the distance to the vector length. By default it's `FALSE`
#'
#' @returns The Hamming distance
#'
#' @export
#' @examples
#' hamming_dist(c(0, 1, 0), c(0, 0, 0))
#' hamming_dist(c(0, 1, 0), c(0, 0, 0), norm = TRUE)
#'
#' ## NAs in any of the vectors are discarded
#' hamming_dist(c(1, NA, 3, 4, 5), c(1, 2, NA, 4, 5))
#' @export
hamming_dist <- function(x, y, norm = FALSE) {
  stopifnot(length(x) == length(y))
  hd <- sum(x != y, na.rm = TRUE)
  `if`(norm, hd / length(x[!is.na(x)]), hd)
}

#' Hamming weight of a binary vector
#'
#' @description
#' The Hamming weight is the number of non null symbols in a vector.
#' For a binary vector, it corresponds to the number of 1s.
#'
#' @details
#' NAs are discarded in the vector by using `na.rm = TRUE` in `sum`.
#'
#' @param v A logical or numeric vector
#' @param norm If `TRUE` (default is `FALSE`) normalize the vector
#'
#' @returns The Hamming weight
#'
#' @export
#' @examples
#' ## Weight of the vector
#' hamming_weight(c(0, 1, 1))
#'
#' ## Normalized weight
#' hamming_weight(c(0, 1, 1), norm = TRUE)
#'
#' ## `NA` are discarded
#' hamming_weight(c(1, 0, NA), norm = TRUE)
hamming_weight <- function(v, norm = FALSE) {
  weight <- sum(v, na.rm = TRUE)
  `if`(norm, weight / length(v[!is.na(v)]), weight)
}

#' Ratio of bits in a binary vector
#'
#' @description
#' The ratio is calculated as the number of 1s minus the number of 0s. A positive ratio indicates that there are more 1s than 0s, while a negative results indicates the opposite. `NA` are not accounted to calculate the length of the vector.
#'
#' @details
#' By using the [hamming_weight][pufr::hamming_weight] function, the ratio can be calculated in the following way.
#' \deqn{Ratio = \frac{HW(v) - (\#v - HW(v))}{\#v} = 2 \cdot HW_{norm}(v) - 1}
#' The operator \eqn{\#v} denotes the number of elements in the vector \eqn{v}.
#'
#' @param v A binary vector
#'
#' @returns The ratio of bits in the binary vector
#'
#' @export
#' @seealso [hamming_weight][pufr::hamming_weight]
#' @examples
#' ## Negative ratio
#' ratio_bits(c(0, 1, 0))
#'
#' ## Positive ratio
#' ratio_bits(c(1, 1, 0))
#'
#' ## `NA` are discarded
#' ratio_bits(c(1, 1, 0, NA, NA))
ratio_bits <- function(v) {
  p_one <- hamming_weight(v, norm = TRUE)
  (2 * p_one) - 1
}

#' @rdname hamming_dist
#' @export
#' @examples
#' c(0, 1, 0) %HD% c(1, 0, 0)
"%HD%" <- function(x, y) hamming_dist(x, y, norm = FALSE)

#' @rdname hamming_dist
#' @export
#' @examples
#' c(0, 1, 0) %NHD% c(1, 0, 0)
"%NHD%" <- function(x, y) hamming_dist(x, y, norm = TRUE)

#' Compare a matrix by pairs of rows
#'
#' @description
#' Each pair of rows is compared using the given function. The pairs are chosen without repetition.
#'
#' @param m Vector of values.
#' @param fn Function that receives two row vectors.
#' @param ... Rest of arguments passed to `fn`.
#'
#' @returns List containing the results of applying the function to each pair of rows.
#' @export
#' @examples
#' #' Compare a matrix by pairs of rows
#' m <- rbits(c(5, 5))
#' res <- compare_pairwise(m, hamming_dist, norm = TRUE)
#' unlist(res)
#' length(res) == (5 * 4 / 2)
#'
#' ## Equivalence to uniqueness
#' res <- compare_pairwise(m, hamming_dist, norm = TRUE)
#' all(uniqueness(m) == unlist(res))
compare_pairwise <- function(m, fn, ...) {
  pairs <- lapply(seq_len(nrow(m) - 1), function(i) {
    len <- length(seq(i + 1, nrow(m)))
    mapply(c, rep(i, len), seq(i + 1, nrow(m)), SIMPLIFY = FALSE)
  })
  pairs <- unlist(pairs, recursive = FALSE)

  lapply(pairs, function(p) fn(m[p[1], ], m[p[2], ], ...))
}

#' Convert a 2D CRP matrix or 3D CRP array to a dataframe
#'
#' @param crps A 2D CRP matrix or 3D CRP array
#'
#' @returns A dataframe with the columns `device`, `challenge` and `response` if `crps` is a 2D matrix. If `crps` is a 3D array, an additional column `sample` is added corresponding to the 3rd dimension.
#'
#' @export
#' @examples
#' ## CRP matrix
#' crps <- rbits(c(2, 5))
#' crps_to_df(crps)
#'
#' ## CRP array
#' crps <- rbits(c(2, 5, 3))
#' crps_to_df(crps)
crps_to_df <- function(crps) {
  mat_to_df <- function(m) {
    reshape2::melt(m) %>%
      rename(device = "Var1", challenge = "Var2", response = "value") %>%
      arrange(device, challenge)
  }

  if (is.matrix(crps)) {
    return(mat_to_df(crps))
  }
  samples <- seq_len(dim(crps)[3])
  arr <- lapply(samples, function(s) mutate(mat_to_df(crps[, , s]), sample = s))
  do.call(rbind, arr)
}

#' Convert a data frame to a CRP matrix
#'
#' The dataframe needs to contain the columns `device`, `challenge` and `response`. The column `sample` is optional and if found, will be used to create a 3D array. Other additional columns are ignored.
#'
#' @param df The dataframe with columns `device`, `challenge`, `response` and optional `sample`
#'
#' @return If the column `sample` is in the data frame, a 3D array. If not, a 2D matrix.
#'
#' @export
#' @examples
#' ## Dataframe without samples
#' df <- expand.grid(device = 1:2, challenge = 1:5)
#' df$response <- rbits(2 * 5)
#' df_to_crps(df)
#'
#' ## Dataframe with samples
#' df <- expand.grid(device = 1:2, challenge = 1:5, sample = 1:3)
#' df$response <- rbits(2 * 5 * 3)
#' df_to_crps(df)
df_to_crps <- function(df) {
  req_names <- c("device", "challenge", "response")
  missing <- !(req_names %in% names(df))
  if (any(missing)) {
    names <- paste0(req_names[missing], collapse = ", ")
    stop(paste("The columns", names, "are missing on the dataframe"))
  }

  devs <- unique(df$device)
  challenges <- unique(df$challenge)
  mat_fn <- function(x) {
    m <- matrix(x, nrow = length(devs), ncol = length(challenges), byrow = TRUE)
    return(m)
  }

  if (!"sample" %in% names(df)) {
    return(mat_fn(df$response))
  }

  arr <- df %>%
    arrange(sample, device, challenge) %>%
    group_by(sample) %>%
    group_map(function(x, ...) mat_fn(x$response)) %>%
    abind::abind(along = 3)

  # TODO: Only assign names if non numeric
  # dimnames(arr) <- list(devs, challenges, unique(df$sample))
  arr
}
